libraries/core_libs/network/src/tarcap/packets_handlers/votes_sync_packet_handler.cpp

```diff
@@ -44,11 +45,37 @@ void VotesSyncPacketHandler::process(const PacketData &packet_data, const std::s
     return;
   }
 
+  std::unordered_map<blk_hash_t, uint64_t> voted_value_round;
   std::vector<std::shared_ptr<Vote>> next_votes;
+
   const auto next_votes_count = packet_data.rlp_.itemCount();
+  //  It is done in separate cycle because we don't need to process this next_votes if some of checks will fail
   for (size_t i = 0; i < next_votes_count; i++) {
     auto next_vote = std::make_shared<Vote>(packet_data.rlp_[i].data().toBytes());
-    const auto next_vote_hash = next_vote->getHash();
+    const auto &next_vote_hash = next_vote->getHash();
+    // votes for the same value should be from one round and step
+    if (voted_value_round.contains(next_vote->getBlockHash())) {
+      // TODO[1964]: I think we should really sync only votes with the same step, but it is not like this in current
+      // design
+      auto previous_votes_round = voted_value_round[next_vote->getBlockHash()];
+      if (previous_votes_round != next_vote->getRound()) {
+        LOG(log_er_) << "Votes sync bundle has votes " << next_vote_hash.abridged() << " with different rounds("
+                     << next_vote->getRound() << ", " << previous_votes_round
+                     << "). The peer may be a malicious player, will be disconnected";
+        disconnect(packet_data.from_node_id_, dev::p2p::UserReason);
+        return;
+      }
+    } else {
+      voted_value_round[next_vote->getBlockHash()] = next_vote->getRound();
+      // We could have votes only for two values. And one of these should be NULL_HASH
+      if (voted_value_round.size() > 2 ||
```

Comment:
I dont follow whole logic around voted_value_round. If we wanna check that there are only two values, we can simply add this check:

```
const auto peer_pbft_period = vote->getPeriod();
const auto peer_pbft_round = vote->getRound();
const auto votes_voted_block = vote->getBlockHash();

if (next_vote->getBlockHash() != votes_voted_block || next_vote->getBlockHash() != NULL_BLOCK_HASH) {
  LOG(log_er_) << "Received next votes bundle with unmatched periods from " << packet_data.from_node_id_
               << ". The peer may be a malicious player, will be disconnected";
  disconnect(packet_data.from_node_id_, dev::p2p::UserReason);
  return;
}
```

Reply:
That is overcomplicated right now for no reason, but original idea was also to check that we have voted for one voted value from the same step